File: src/2360. Longest Cycle in a Graph.js
/**
 * @param {number[]} edges
 * @return {number}
 */
export function longestCycle(edges) {
  const n = edges.length
  const visited = Array(n).fill(false)
  let maxCyclePath = -1
  for (let i = 0; i < n; i++) {
    if (visited[i]) {
      continue
    }
    const paths = new Map()
    let distance = 0
    for (let j = i; j !== -1; j = edges[j]) {
      if (paths.has(j)) {
        maxCyclePath = Math.max(maxCyclePath, distance - paths.get(j))
        break
      }
      if (visited[j]) {
        break
      }
      visited[j] = true
      paths.set(j, distance)
      distance++
    }
  }

  return maxCyclePath
}

File: src/2360. Longest Cycle in a Graph.cpp
class Solution {
public:
    int longestCycle(vector<int>& edges) {
        const int n = edges.size();
        vector<bool> visited(n, false);
        int maxCyclePath = -1;
        for (int i = 0; i < edges.size(); ++i) {
            if (visited[i]) {
                continue;
            }
            unordered_map<int, int> paths;
            int distance = 0;
            for (int j = i; j != -1; j = edges[j]) {
                if (paths.find(j) != paths.end()) {
                    maxCyclePath = max(maxCyclePath, distance - paths[j]);
                    break;
                }
                if (visited[j]) {
                    break;
                }
                visited[j] = true;
                paths[j] = distance;
                distance++;
            }
        }
        return maxCyclePath;
    }
};